### PR TITLE
Resolving NEOS interface issues

### LINKED ISSUE #1201: Issues with NEOS solvers and solution output
I'm with the NEOS team and I'm trying to figure out some issues with Pyomo+NEOS.

It seems that 
`pyomo solve --solver=<solver> --solver-manager=neos model.py`
works only if `<solver>` has an explicit plugin (e.g., any of the solvers in pyomo/solvers/plugins/solvers).  This is mentioned in #910.  

The following seems to work for any NEOS-valid solver:
```
manager = SolverManagerFactory('neos')
results = manager.solve(model,opt='<solver>')
```

Is there a bug with the first method?  And if not, I assume that the second method is the way to go?  I dug around a little bit and it looks like the first method tries to create `SolverFactory(<solver>)` but this results in an unknown solver for some of the NEOS solvers.

Furthermore, in either case, no solution seems to be returned.  I don't see any error messages, but printing the results always shows 
```
Solution: 
- number of solutions: 0
  number of solutions displayed: 0
```
I'm not sure what Pyomo is expecting from NEOS, but if there's anything we can do on our end to fix this, let me know.

### PR DESCRIPTION
**Pull request by whart222**
## Summary/Motivation:

This PR resolves issues highlighted in #1201 .  There remain some solver interface issues that relate to limitations of NEOS solvers (e.g. that fail on legal problems that Pyomo sends), but otherwise I think the issues in #1201 are resolved.

## Changes proposed in this PR:
- New tests for the NEOS interface
- Changed NEOS tests to category 'neos'.
- Fixes an issue using the neos solver manager from the Pyomo commandline.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by whart222** [comment]
NOTE:  I changed the NEOS tests to use the 'neos' category.  This should disable all NEOS tests in our nightly build.  This isn't what we want, though.  We want just one test to use the 'neos' category.

I'm not sure how to revise the testing script to define this test appropriately.

**Comment by mrmundt** [comment]
> NOTE: I changed the NEOS tests to use the 'neos' category. This should disable all NEOS tests in our nightly build. This isn't what we want, though. We want just one test to use the 'neos' category.
> 
> I'm not sure how to revise the testing script to define this test appropriately.

@jsiirola can correct me if I'm wrong, but I believe you can put more than one category. Is there any problem with putting `unittest.category('neos', 'nightly')`?

File: pyomo/neos/tests/test_neos.py [hunk 1]
```diff
@@ -90,6 +88,177 @@ def test_kestrel_plugin(self):
...
+        kestrel = kestrelAMPL()
+        tmp = [tuple(name.split(':')) for name in kestrel.solvers()]
+        amplsolvers = set(v[0].lower() for v in tmp if v[1]=='AMPL')
+
+        doc = pyomo.neos.doc
+        dockeys = set(doc.keys())
+
+        self.assertEqual(amplsolvers, dockeys)
+
+        #gamssolvers = set(v[0].lower() for v in tmp if v[1]=='GAMS')
```
Comment (mrmundt) [inline pyomo/neos/tests/test_neos.py:101]: These lines are commented out - do we still need them?

File: pyomo/neos/tests/test_neos.py [hunk 2]
```diff
@@ -90,6 +88,177 @@ def test_kestrel_plugin(self):
...
+            '--save-results=%s' % results,
+            '-c'
+            ]
+        try:
+            output = main.run(args)
+            self.assertEqual(output.errorcode, 0)
+
+            with open(results) as FILE:
+                data = json.load(FILE)
+            #print(json.dumps(data, indent=4))
```
Comment (mrmundt) [inline pyomo/neos/tests/test_neos.py:235]: Again - commented out. Is this needed?

File: pyomo/scripting/plugins/solve.py [hunk 3]
```diff
@@ -202,6 +210,9 @@ def solve_exec(args, unparsed):
 
     if config is None:
         raise RuntimeError("Failed to create config object")
+    config.solvers[0].solver_name = solver
+    config.solvers[0].manager = solver_manager
+    #config.display()
```
Comment (mrmundt) [inline pyomo/scripting/plugins/solve.py:215]: Commented out. Is this needed?

File: pyomo/neos/tests/test_neos.py [hunk 4]
```diff
@@ -90,6 +88,177 @@ def test_kestrel_plugin(self):
...
+            if os.path.exists(results):
+                os.remove(results)
+
+
+@unittest.category('neos')
+@unittest.skipIf(not neos_available, "Cannot make connection to NEOS server")
+class TestSolvers_cmd_max(TestSolvers_cmd_min):
+
+    filename = 't1.py'
+    objective = 1
```
Comment (mrmundt) [inline pyomo/neos/tests/test_neos.py:260]: Is this test complete? It doesn't have anything besides a filename and objective.

**Review by mrmundt — CHANGES_REQUESTED** [review]

File: pyomo/neos/tests/test_neos.py [hunk 5]
```diff
@@ -90,6 +88,177 @@ def test_kestrel_plugin(self):
...
+    #def test_minto(self):
+    #    self._run('minto')
+
+    def test_mosek(self):
+        self._run('mosek')
+
+    # This solver doesn't give the same answer for
+    # min f(x) and max -f(x)
+    #def test_ooqp(self):
+    #    self._run('ooqp')
```
Comment (jsiirola) [inline pyomo/neos/tests/test_neos.py:182]: If we add class-level attributes that indicates the current sense, then we can toggle the expected results here (through optional argument to `_run()`?) and restore this test.

File: pyomo/neos/tests/test_neos.py [hunk 6]
```diff
@@ -90,6 +88,177 @@ def test_kestrel_plugin(self):
...
+    # problems
+    #def test_path(self):
+    #    self._run('path')
+
+    def test_snopt(self):
+        self._run('snopt')
+
+    # This test only handles minimization problems
+    #def test_raposa(self):
+    #    self._run('raposa')
```
Comment (jsiirola) [inline pyomo/neos/tests/test_neos.py:194]: With a class level attribute that indicates the sense, we could skip this test for maximization cases.

File: pyomo/scripting/plugins/solve.py [hunk 7]
```diff
@@ -29,6 +29,10 @@ def create_parser(parser=None):
         action='store',
         dest='solver',
         default=None)
+    parser.add_argument('--solver-manager',
+        action='store',
+        dest='solver_manager',
+        default='serial')
```
Comment (jsiirola) [inline pyomo/scripting/plugins/solve.py:35]: This makes me nervous: it implies users have more options available to them than they do.  What happens if the user specifies something like `async`?

Does this also represent a change in the way users are supposed to interact with NEOS from the command line?  Is there documentation that needs to be updated?

**Review by jsiirola — CHANGES_REQUESTED** [review]
Several questions and a couple requested changes.

File: pyomo/neos/tests/test_neos.py (hunk 1 above)
Comment (whart222) [reply pyomo/neos/tests/test_neos.py:101]: Please leave these to document the test change we should consider later.

File: pyomo/neos/tests/test_neos.py (hunk 2 above)
Comment (whart222) [reply pyomo/neos/tests/test_neos.py:235]: No.  Delete

File: pyomo/scripting/plugins/solve.py (hunk 3 above)
Comment (whart222) [reply pyomo/scripting/plugins/solve.py:215]: No.  Delete.

File: pyomo/neos/tests/test_neos.py (hunk 4 above)
Comment (whart222) [reply pyomo/neos/tests/test_neos.py:260]: Correct.

File: pyomo/neos/plugins/kestrel_plugin.py [hunk 8]
```diff
@@ -48,6 +48,11 @@ def _neos_error(msg, results, current_message):
     'neos', doc="Asynchronously execute solvers on the NEOS server")
 class SolverManager_NEOS(AsynchronousSolverManager):
 
+    def __del__(self):
+        # This is a hack to force the socket to close
+        if self.kestrel.neos is not None:
+            self.kestrel.neos.__exit__()
```
[the remark this answers is not in the document]
Comment (whart222) [reply pyomo/neos/plugins/kestrel_plugin.py:54]: I think I added the __del__ method in kestrelAMPL first, but then ran into a case where I needed this.  You can try deleting this, but I think you'll see this case as well if you run test_neos.py by hand.

File: pyomo/neos/tests/test_neos.py [hunk 9]
```diff
@@ -90,6 +88,177 @@ def test_kestrel_plugin(self):
...
+                    data['Solution'][1]['Objective']['o']['Value'], self.objective)
+        finally:
+            cleanup()
+            if os.path.exists(results):
+                os.remove(results)
+
+
+@unittest.category('neos')
+@unittest.skipIf(not neos_available, "Cannot make connection to NEOS server")
+class TestSolvers_cmd_max(TestSolvers_cmd_min):
```
[the remark this answers is not in the document]
Comment (whart222) [reply pyomo/neos/tests/test_neos.py:257]: Yes, I agree.  Let's deleted the scripting_max tests.

File: pyomo/scripting/plugins/solve.py (hunk 7 above)
Comment (jsiirola) [reply pyomo/scripting/plugins/solve.py:35]: Ahh - after discussing this with @whart222, this is only promoting an option that was previously available through the specific solver interface up to the subcommand level, and does not represent a change in the user interface.

File: pyomo/neos/kestrel.py [hunk 10]
```diff
@@ -121,11 +121,25 @@ def make_connection(self, host):
...
+        # This is a hack to force the socket to close.  When running
+        # tests under unittest, unittest will report any underlying
+        # socket resources that have not been closed as warnings.  This
+        # will cause the socket to be closed when the kestrelAMPL object
+        # falls out of scope and will suppress the warnings from
+        # unittest.
+        #
+        # Note that this is only to suppress warnings, as __del__ is not
+        # guaranteed to be called (especially for any onjects that still
+        # exist when the python process terminates)
```
Comment (blnicho) [inline pyomo/neos/kestrel.py:139]: Typo: objects, **P**ython

**Review by blnicho — APPROVED** [review]
Found a couple typos but otherwise this looks fine

**Review by mrmundt — APPROVED** [review]

**Review by jsiirola — APPROVED** [review]

**Pull request merged**